### ISSUE TITLE
Matching on Map and Set

### ISSUE THREAD

**Issue by ttuegel** [runtimeverification/k#1134]
This documents a conversation today between @ttuegel, @dwightguth, and @sskeirik. We discussed the Haskell backend's solution to kframework/kore#672 and how to reconcile that solution with the behavior of other backends and with the intended semantics of existing K definitions. Refer to that issue to understand the rationale for this solution; these remarks will only document the facts of that solution and our path forward.

This proposal applies equally to Maps and Sets, but we will only explicitly discuss Maps. Note that a Set is a Map where all keys take values in the Unit sort (the sort with one inhabitant).

There are two operations on Map related to the union of Maps, which we will call `concatMap` and `unionMap`, with the following semantics:

```
syntax Map ::= concatMap(Map, Map) [function]
// concatMap is associative and commutative, and its unit is .Map
// concatMap is #False (\bottom) if its arguments have keys in common:
rule concatMap(K |-> _, K |-> _) => #False
// otherwise, concatMap is the "union" in the way you would expect

syntax Map ::= unionMap(Map, Map) [function]
// unionMap is associative and commutative, and its unit is .Map
// unionMap is #False (\bottom) if its arguments' common keys have different values:
rule unionMap(K |-> V1, K |-> V2) => #if V1 ==K V2 #then K |-> V1 #else #False
// otherwise, unionMap is the "union" in the way you would expect
```

Note that it is simple to implement `unionMap` in terms of `concatMap`, but the reverse is not so simple.

The Haskell backend gives the juxtaposition of Maps (`_Map_`) the semantics of `concatMap`. The Java and LLVM backends give the juxtaposition of Maps the semantics of `unionMap` when it appears on the right-hand side of a rule, and the semantics of `concatMap` when it appears on the left-hand side, although in both cases, `_Map_` is translated into Kore as the same symbol.

To harmonize implementations, we suggest that `_Map_` should be translated into Kore differently on the left- and right-hand sides of rules. On the left-hand side it will be translated to a `concatMap` symbol, and on the right-hand side it will be translated to a `unionMap` symbol. In addition to changing the frontend, this will require implementing the `concatMap` function in the LLVM backend and the `unionMap` function in the Haskell backend. (In the latter case, we again note that it is possible to define `unionMap` in K.)

**Comment by dwightguth** [comment]
I'm going to flesh out a little bit more the changes that will be needed to the frontend and the llvm backend. I assume that Tom is much better suited to fleshing out the changes that need to be made to the Haskell backend.

Frontend:

* [ ] We need to add a new concatMap and concatSet symbols to the MAP and SET modules in domains.k. These will be given the MAP.concat and SET.concat hooks, and the existing productions with those hooks will be given the MAP.union and SET.union hooks. 
* [ ] We probably also need to add an attribute to the `_Set_` and `_Map_` symbols that will tell the frontend where to look for the corresponding nilpotent operator.
* [ ] As a transformer implemented on the level of K immediately prior to ModuleToKORE, we need to transform every instance of `_Set_` and `_Map_` on the lhs of rules into the corresponding nilpotent operator.
* [ ] We will need to audit every use of MAP.concat and SET.concat in the codebase and either replace it with the corresponding idempotent hook, or leave it alone.

LLVM Backend:

* [ ] the functions hook_MAP_concat and hook_SET_concat should be renamed hook_MAP_union and hook_SET_union.
* [ ] the code that deals with pattern matching should be modified so that it interprets both MAP.concat and MAP.union identically on the lhs. the same is true of SET. This is for backwards compatibility.
* [ ] the hook_MAP_concat and hook_SET_concat functions will need to be written to match their intended semantics.
* [ ] Minor changes may need to be made to the test suite to keep it passing.

**Comment by nishantjr** [comment]
Re the second point on the LLVM Backend list, why would `*.union` ever occur on the LHS of a rule?
Wouldn't we only have `_Map_` or `_Set_`, allowing us to choose the correct one at translation time?

**Comment by ttuegel** [comment]
Supporting the concrete case in the Haskell backend would not take much work:

- [ ] Fix kframework/kore#1183
- [ ] Implement `MAP.union` and `SET.union` symbolically in `domains.k`.

Fully supporting the symbolic case presents some difficulty. In the concrete case, every union can be evaluated to a concat, but not so in the symbolic case. At least, not without branching:

- [ ] Allow branching on designated functions, such as `unionMap` and `unionSet`.
